### DISCUSSION TITLE
setAtom function is permissive and allows extra props ?

### DISCUSSION THREAD

**Discussion by ayoubqrt** [pmndrs/jotai#1796] [Q&A]
Hi everyone,

It's my first time using Jotai and I love the package ! 
Unfortunately I am faced to a bug/problem. When I want to set the value of an atom. I see that that the setAtom function is permissive and allow extra props that are not declared in the atom.

Without the previous state parameter, typescript screams at me:

`setPlayerAtom({
        ...playerAtom,
        foo: 23,
});
`

But with the previous state, it works and typescript is happy:

`setPlayerAtom((player) => ({
        ...player,
        foo: 23,
}))`

You can see the reproduction in this typescript playground: 

https://www.typescriptlang.org/play?ssl=21&ssc=18&pln=20&pc=1#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQokscA3ijCQDRwCuqWAQQ6kAvnCIlyAKwgxkwWnQYA7GFigE8WOAAUANsgCe6oZJZ04cEH2C4AKsBh6sALjioYUYMoDm9EUq4EMoecGAGxlAAyhyUcAC87CQAPPpGJsIAfAAU5pbWqLYOTq7kZJx0IgCU9HRBIfAAsoYYyMoAJiQAwiSQyliqCXDZVQmZrBaWcPWhANrh6VCmINz8MGmRywC6Q3yCwtkLkTHQWDVKU3AA9FdwAHKycADu0ADW3j5wAEZYuMh7cAABgQIBBAXBgOhlI9vLoIhlJDBDGAsJNLGsNgiQLk0Zc4AA6QlHLEVPFTEEQNwAJgAzKSptVapcbnAAOpvD5wFzZXEY+FLA6HfmjeLjHFkyyE-HEqD0skU6l03HVKoXKaUGA8KDKODJdrAABumWSV31Rv8QA

Can I open an issue ?

**Comment by dai-shi** [comment]
Can we assume it's how TS works?
https://tsplay.dev/w6xqym
```ts
    // Working too
    const nextValue = {
      ...playerAtom,
      foo: 23,
    }
    setPlayerAtom(nextValue);
```

I don't think there's a reasonably straightforward solution.
If you have any suggestion, let's discuss it here.

**Reply by ayoubqrt** [reply]
I'm doing researches and yes I assume that TS is working like this with structural typing:
https://github.com/microsoft/TypeScript/wiki/FAQ#what-is-structural-typing

Now, I don't know if it's Jotai's responsibility to do this but the way to fix the problem is to use a custom type that will be strict on the extra props.

**Comment by wenq1** [comment]
ill suggest an interesting read here:
https://github.com/microsoft/TypeScript/issues/12936

Multiple solutions to your EPC requirement exist, and not all of them behave the same. I suggest you try a few to see which one fits your need.
